torch/utils/_sympy/value_ranges.py

```diff
@@ -96,40 +109,42 @@ def tighten(self, other) -> "ValueRanges":
         return self & other
 
     # Intersection
-    def __and__(self, other) -> "ValueRanges":
+    def __and__(self: ValueRanges[T], other: ValueRanges[T]) -> ValueRanges[T]:
         if other == ValueRanges.unknown():
             return self
         if self == ValueRanges.unknown():
             return other
         assert self.is_bool == other.is_bool, (self, other)
-        if self.is_bool:
-            range = ValueRanges(sympy.Or(self.lower, other.lower), sympy.And(self.upper, other.upper))
+        if vr_is_bool(self):
+            return cast(ValueRanges[T], ValueRanges(sympy.Or(self.lower, other.lower), sympy.And(self.upper, other.upper)))
+        elif vr_is_expr(self):
+            return cast(ValueRanges[T], ValueRanges(sympy.Max(self.lower, other.lower), sympy.Min(self.upper, other.upper)))
         else:
-            range = ValueRanges(sympy.Max(self.lower, other.lower), sympy.Min(self.upper, other.upper))
-        return range
+            raise AssertionError("impossible")
 
     # Union
     def __or__(self, other) -> "ValueRanges":
         if ValueRanges.unknown() in (self, other):
             return ValueRanges.unknown()
         assert self.is_bool == other.is_bool, (self, other)
-        if self.is_bool:
-            range = ValueRanges(sympy.And(self.lower, other.lower), sympy.Or(self.upper, other.upper))
+        if vr_is_bool(self):
+            return cast(ValueRanges[T], ValueRanges(sympy.And(self.lower, other.lower), sympy.Or(self.upper, other.upper)))
+        elif vr_is_expr(self):
+            return cast(ValueRanges[T], ValueRanges(sympy.Min(self.lower, other.lower), sympy.Max(self.upper, other.upper)))
         else:
-            range = ValueRanges(sympy.Min(self.lower, other.lower), sympy.Max(self.upper, other.upper))
-        return range
+            raise AssertionError("impossible")
```

Comment:
```suggestion
            typing_extensions.assert_never(self)
```
Is how you are suppose to handle unreachable code with typecheckers.

Reply:
Can't do this, mypy won't be able to prove that the code is unreachable because it doesn't know that vr_is_bool and vr_is_expr covers all cases